howto/stream/oob-data.md

```diff
@@ -106,15 +106,131 @@ A trivial example to simulate the data transmission between an Anbox container a
+        } catch (ClassNotFoundException e) {
+            e.printStackTrace();
+        }
+        return service;
+    }
+    ```
+
+-  If you ship the Android application inside of the AOSP source tree and [build](https://source.android.com/docs/setup/build/building) it from there, you can use Android's hidden API to access the service.
+
+    ```
+    IBinder getDataProxyService() {
+        return ServiceManager.getService("org.anbox.webrtc.IDataProxyService");
+    }
+    ```
+
+#### Connect the data channel
+
+To fetch the file descriptor that refers to one data channel, send a request to the data proxy service through a binder transaction:
+
+```
+ParcelFileDescriptor mFd = null;
+String channel = "foo";   // denotes data channel name
+Parcel data = Parcel.obtain();
+Parcel reply = Parcel.obtain();
+try {
+    data.writeInterfaceToken("org.anbox.webrtc.IDataProxyService@1.0");
+    data.writeString(channel);
+    mService.transact(TRANSACTION_connect, data, reply, 0);
+    mFd = reply.readFileDescriptor();
+    if (mFd.getFd() < 0) {
+        Log.e(TAG, "Invalid file descriptor");
+        return;
+    }
+    ...
+    ...
+} catch (RemoteException ex) {
+    Log.e(TAG, "Failed to connect data channel '" +  channel + "': " + ex.getMessage());
+} finally {
+    data.recycle();
+    reply.recycle();
+}
+```
+
+#### Receive data from the Anbox WebRTC platform
+
+Once the valid file descriptor is returned, launch an asynchronous task to read data from the Anbox WebRTC platform:
+
+```
+public class DataReadTask extends AsyncTask<Void, Void, Void> {
+    ...
+    ...
+    @Override
+    protected Void doInBackground(Void... parameters) {
+        try (InputStream in = new ParcelFileDescriptor.AutoCloseInputStream(mFd)) {
+            byte[] data = new byte[1024];
+            while (!isCancelled()) {
+                int read_size = in.read(data);
+                if (read_size < 0) {
+                    Log.e(TAG, "Failed to read data");
+                    break;
+                } else if (read_size == 0) {
+                    // EOF reached
+                    break;
+                }
+
+                byte [] readBytes = Arrays.copyOfRange(data, 0, read_size);
+                ...
+                ...
+            }
+        } catch (IOException ex) {
+            if (!isCancelled())
+                Log.e(TAG, "Failed to read data: " + ex);
+        }
+
+        return null;
+    }
+}
+```
+
+#### Send data to the Anbox WebRTC platform
+
+To send data to the Anbox WebRTC platform through the file descriptor:
+
+```
+OutputStream ostream = new FileOutputStream(mFd.getFileDescriptor());
+try {
+    ostream.write(data.getBytes(), 0, data.length());
+} catch (IOException ex) {
+    Log.i(TAG, "Failed to write data: " + ex.getMessage());
+    ex.printStackTrace();
+}
+```
+
+For a complete Android example, see the [out_of_band_v2](https://github.com/anbox-cloud/anbox-streaming-sdk/tree/master/examples/android/out_of_band_v2) project.
```

Comment:
The link doesn't exist yet - I guess the project will be merged before we publish?